### ISSUE TITLE
nil pointer dereference on SIGINT

### ISSUE THREAD

**Issue by heyimalex** [mheese/journalbeat#71]
Running journalbeat from the command line for testing. Rarely (though more than once) pressing ctrl-c I get: 

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x46800c]

goroutine 43 [running]:
panic(0x8eb2c0, 0xc4200100c0)
        /usr/local/go/src/runtime/panic.go:500 +0x1a1
github.com/mheese/journalbeat/beater.(*Journalbeat).managePendingQueueLoop(0xc4200e4000)
        /vagrant/src/github.com/mheese/journalbeat/beater/state.go:138 +0x5fc
created by github.com/mheese/journalbeat/beater.(*Journalbeat).Run
        /vagrant/src/github.com/mheese/journalbeat/beater/journalbeat.go:140 +0x116
```

Offending line is [here](https://github.com/mheese/journalbeat/blob/master/beater/state.go#L138).

**Comment by heyimalex** [comment]
You probably already know this, but something kind of surprising/non-intuitive about go's select is that if multiple cases are available at the same time the runtime will choose which branch to take randomly. Maybe could be the reason it's pretty rare? A race condition + a non-determistic choice.

**Comment by heyimalex** [comment]
Ok, here goes.

- `Stop()` is called, which
- calls `close(done)`, which
- makes `range journal.Follow(jb.journal, jb.done)` complete, which
- ends `Run()`, so the defered func at the top execute, which
- calls `close(pending)`

So, _technically_ `close(done)` gets called before `close(pending)`, but not in any way that will guarantee the order in that select. The select runs and pending is both empty and closed, so if it's randomly chosen from all of the unblocked cases it will return the zero value for `*eventReference`, which is where we're getting the nil pointer exception.

**Comment by nicorevin** [comment]
Yes, you are right. closing pending and completed queues was put in the wrong place. I will move them and fix this. It is a pity that `-race` flag does not detect it. Will try to find out whether this case has been covered by any tool. If not will raise it with the Go team directly.

**Comment by nicorevin** [comment]
will make a PR soon in regard to this issue.

**Comment by nicorevin** [comment]
I was able to test it now. The problem is that having done that change I get this rather frequently:
```
panic: send on closed channel

goroutine 1 [running]:
github.com/mheese/journalbeat/beater.(*eventSignal).Completed(0xc420286460)
```

**Comment by nicorevin** [comment]
I hadn't understood the problem fully and made a mistake. 

Closing both channels in the deferred function on return from `Run()` is fine. 

But I did not realize that we can read a zero value from a closed channel in a select statement (that's why I would rather use a range loop to read from channels...)

Fixed it.
